### PR TITLE
YoloX integration

### PR DESCRIPTION
**Pull request by picsalex**
This PR brings the new yolox model that comes from here: https://github.com/Megvii-BaseDetection/YOLOX.

The model brings several changes with some paradigms:
- It only allows `train`, `test` and `val` dataset
- The `val` dataset is used to perform evaluations at the end of each epoch
- The `test` dataset is used to perform the evaluation at the end of the training
- Everything located inside `yolox-detection/experiment/YOLOX` SHOULD NOT be review, as it comes directly from the original yolox repository.
- [x] Only feature that remains to be developed is handling 1 or 2 datasets (for now only a training with exactly 3 datasets is supported)
-> Warning ⚠️ : Some files have been linted by `black` during the development phase, another PR that integrate `black` as a pre-commit step should be included

### CODE REVIEW

File: yolox-detection/experiment/main.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    yolox_predictor=yolox_predictor,
+    type_formatter=type_formatter,
+    experiment=experiment,
+    dataset=test_ds,
+)
+
+# 9 - Export the model to ONNX
+model = replace_module(model, torch.nn.SiLU, SiLU)
+model.head.decode_in_inference = False
+dummy_input = torch.randn(1, 3, image_size, image_size)
```
Comment (SoniaGrh) [inline yolox-detection/experiment/main.py:148]: Il ne faudrait pas remplacer le 1 par la batch_size ?

File: yolox-detection/experiment/main.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+# 6B - Get model architecture
+exp = get_exp_by_name(args)
+exp.merge(args.opts)
+check_exp_value(exp)
+
+if not args.experiment_name:
+    args.experiment_name = exp.exp_name
+
+num_gpu = get_num_devices() if args.devices is None else args.devices
```
Comment (SoniaGrh) [inline yolox-detection/experiment/main.py:108]: le args.devices n'est pas toujours à None ici ?

File: yolox-detection/experiment/main.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+# 7 - Prepare the model for inference and export
+model = exp.get_model()
+
+file_name = os.path.join(exp.output_dir, args.experiment_name)
+ckpt_file = os.path.join(file_name, "best_ckpt.pth")
+
+if not os.path.isfile(ckpt_file):
+    ckpt_file = os.path.join(file_name, "latest_ckpt.pth")
+
+ckpt = torch.load(ckpt_file, map_location="cpu")
```
Comment (SoniaGrh) [inline yolox-detection/experiment/main.py:123]: pourquoi ne pas donner la map_location en fonction du num_gpu ?

File: yolox-detection/experiment/main.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+model.head.decode_in_inference = False
+dummy_input = torch.randn(1, 3, image_size, image_size)
+model_path = os.path.join(exp.output_dir, args.experiment_name, "best.onnx")
+
+torch.onnx.export(
+    model,
+    dummy_input,
+    model_path,
+)
+experiment.store("model-latest", model_path)
```
Comment (SoniaGrh) [inline yolox-detection/experiment/main.py:156]: tu pourrais peut être store aussi le modèle sous forme .pth pour le donner en entrée si tu veux réentrainer le modèle :
image

File: yolox-detection/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,14 @@
+FROM picsellia/cuda:11.7.1-cudnn8-ubuntu20.04
+
+COPY ./yolox-detection/requirements.txt .
+
+ARG REBUILD_ALL
+RUN python3.10 -m pip install -r requirements.txt --no-cache-dir
```
Comment (SoniaGrh) [inline yolox-detection/Dockerfile:6]: si ton environnement marche avec poetry, tu pourras remplacer l'installation avec pip par celle avec poetry

**Review by PN-picsell — APPROVED** [review]

**Pull request merged**